README
Rhythmbox MicroBlogger

DESCRIPTION
  A simple plugin to send current music information to a micro blogging service

AUTHORS
  Ali Vakilzade <ali.vakilzade [--at--] gmail [--dot--] com>

  GetGlue.com by Matthew Gregg <mcg {at} braintube.com>

LINKS

  page on freashmeat (please vote)  : http://goo.gl/Uu8yM
  page on ohloh (please vote)       : http://goo.gl/YLCI1
  page on github (report bugs here) : http://goo.gl/8eDjx

  group on identi.ca (please join)  : http://goo.gl/Lnrmv

  screen shot                       : http://goo.gl/CBehH


INSTALL
  0-
    from git (last version)
      mkdir -p ~/.gnome2/rhythmbox/plugins
      cd ~/.gnome2/rhythmbox/plugins
      git clone git://github.com/aliva/rhythmbox-microblogger.git

    from Repositories
      Arch Linux                    : http://goo.gl/HwBvX (last version from git)
                                    : http://goo.gl/pA7aU (stable package)
      Frugalware Linux              : http://goo.gl/DFFu3

    from archive:
      tar -xvf rhythmbox-microblogger-[xxx].tar.gz
      mkdir -p ~/.gnome2/rhythmbox/plugins
      cp rhythmbox-microblogger ~/.gnome2/rhythmbox/plugins -Rf

  1- run/restart Rhythmbox active this plugin 'microblogger' and add accounts!

  2- now a button will appear in rhythmbox toolbar. use that for sending notice

  3- If you had problems. start rhythmbox with this command (debug mode)
        rhythmbox -D rhythmbox-microblogger
     and check output.

IMPORTANT
  * when I change the name of directory from 'rhythmbox-micoblogger' to something else
    plugin can not find images. so can not be activated.

  * You need these packages for this plugin (package names in Fedora)
    - Rhythmbox
    - python
    - gnome-python2-gconf
    - python-notify
    -- On other distros you may need python extra packages such as urllib, hashlib ....

COPYING
  microblogger is distributed under the GPL Version 3.




ChangeLog
0.5.2, 2011-Jan-18:
    * GetGlue.com added by Matthew
	* proxy added
	* now you can turn off progress bar
	* Escape key closes message box
	* code impproved
	* new bugs added!


0.5.1, 2010-Dec-5:
    * checks plugin updates
    * use tags in template 
        - Eg: add #{artist} in template and it will replaced with #music_artist instead of music artist.

0.5.0, 2010-Dec-1:
    * oauth module changed from tweepy to python-oauth2
    * code was written from base
    * interface improved
    * uses libnotify
    * in-time text replace
        - if you enter {title} in message. it will changed to music title. check configure window for more.

0.4.2, 2010-Sep-12:
    * little improves

0.4.0, 2010-Sep-09:
    * there is noting. I don't remember why I changed tag. lol (maybe some unpushed/uncommited changes??)

0.3.1, 2010-Sep-09:
    * oauth support added using tweepy module.
    * twitter works!

0.2.1, 2010-Aug-31:
    * closes url after sending notice

0.2.0, 2010-Aug-31:
    * support for thread added (ui won't freeze any more)

0.1.1, 2010-Aug-31:
    * conf-dialog improved

0.1.0, 2010-Aug-29:
    * First release 


rbmbRequest.py
#! /usr/bin/python2
# -*- coding: utf8 -*-
#
# Rhythmbox-Microblogger - <http://github.com/aliva/Rhythmbox-Microblogger>
# Copyright (C) 2010 Ali Vakilzade <ali.vakilzade in Gmail>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import base64
import hashlib
import hmac
import libxml2
import pynotify
import random
import time
import urlparse
import urllib
import urllib2
import webbrowser
import oauth2 as oauth

IDENTICA={
    'key'   :'NzljNWU2MDFjNmQzMTU0ZDRhMTkwMTRmZmI1MWU2Zjk=',
    'secret':'YzgyYmJiZDg3NWVlYmM2ZWZkODA3OTEwYjg3M2VhMDk=',
    'request_token':'https://identi.ca/api/oauth/request_token',
    'access_token' :'https://identi.ca/api/oauth/access_token',
    'authorization':'https://identi.ca/api/oauth/authorize',
    'post':'https://identi.ca/api/statuses/update.json',
    'call_back':'oob',
    'maxlen':140,
}

TWITTER={
    'key'   :'NlFmM0JtVmpETk1UOUlYek9oa1E0Zw==',
    'secret':'QVd6SnBldWNvM0dPU0pXRlpGcGJpeXlJOGNlSnRWb1k4TmRZdHQzVVpn',
    'request_token':'https://twitter.com/oauth/request_token',
    'access_token' :'https://twitter.com/oauth/access_token',
    'authorization':'https://twitter.com/oauth/authorize',
    'post': 'https://twitter.com/statuses/update.json',
    'call_back':None,
    'maxlen':140,
}

GETGLUE={
    'key'   :'YzAzMGU0NGUxZmJmNzllZWU4Zjg2YjA0ZDAzNGYxZWI=',
    'secret':'NWEyZjc5M2FhMGVlNTBhNWM5MWEwN2VhYzhhYjBmNmI=',
    'request_token':'http://api.getglue.com/oauth/request_token',
    'access_token' :'http://api.getglue.com/oauth/access_token',
    'authorization':'http://getglue.com/oauth/authorize',
    'post': 'http://api.getglue.com/v2/user/addCheckin?',
    'call_back':None,
    'maxlen':140,
}

class AddAccountRequest():
    def __init__(self):
        self.api=None
        self.type=None
        
        self.request_token=None
        self.alias=None

        
    def __del__(self):
        pass

        
    def request_set_type(self, t):
        self.type=t
        
    def authorize(self, *args):
        set_hint, button, assistant, page, proxy_info=args
        
        self.pin=''

        button.set_sensitive(False)
           
        if self.type=='identica':
            api=IDENTICA
        elif self.type=='twitter':
            api=TWITTER
        elif self.type=='getglue':
            api=GETGLUE

        self.api=api            
        
        self.consumer=oauth.Consumer(decode(api['key']), decode(api['secret']))
        client=oauth.Client(self.consumer, proxy_info=proxy_info)
           
        try:     
            resp, content = client.request(api['request_token'], "GET",
                                           call_back=api['call_back'])
        except Exception as err:
            set_hint('ERR: %s' % err)
            button.set_sensitive(True)
            return
        
        if resp['status'] != '200':
            set_hint('ERR: %s' % (resp['status']))
            button.set_sensitive(True)
            return
        
        self.request_token=dict(urlparse.parse_qsl(content))
        
        
        url = "%s?oauth_token=%s" % (self.api['authorization'] , self.request_token['oauth_token'])
        set_hint('Opening your web browser')
        print url
        webbrowser.open_new(url)
        
        page[4]=True
        assistant.set_page_complete(page[0], page[4])

    def exchange(self, *args):
        set_hint, button, assistant, page, proxy_info=args
        
        button.set_sensitive(False)
        token=oauth.Token(self.request_token['oauth_token'],
                          self.request_token['oauth_token_secret'])
        token.set_verifier(self.pin)
        client = oauth.Client(self.consumer, token, proxy_info=proxy_info)
        
        try:
            resp, content = client.request(self.api['access_token'], "POST")
        except Exception as err:
            set_hint('ERR: %s' % err)
            button.set_sensitive(True)
            return
        
        if resp['status'] != '200':
            set_hint('ERR: %s\n%s' % (resp['status'], content))
            button.set_sensitive(True)
            return
        
        access_token = dict(urlparse.parse_qsl(content))

        self.access_token = access_token['oauth_token']
        self.access_token_secret = access_token['oauth_token_secret']
        
        set_hint('Done!\nnow you can forward to next page')
        
        page[4]=True
        assistant.set_page_complete(page[0], page[4])
    
    def save_account(self, mb):
        mb.settings.add_account(
                type=self.type,
                alias=self.alias,
                token=encode(self.access_token),
                token_secret=encode(self.access_token_secret),
                url='',
                maxlen=self.api['maxlen'])

class Post:
    def __init__(self, mb):
        self.mb=mb
        
    def post(self, *args):
        ui, self.artist, self.title, self.album, alias, proxy_info=args
        
        conf=self.mb.get_conf('a', alias)
        
        if conf['type']=='twitter':
            api=TWITTER
        elif conf['type']=='identica':
            api=IDENTICA
        elif conf['type']=='getglue':
            api=GETGLUE
        
        get=ui.get_object
        self.get=get
        
        w=get('general')
        w.set_sensitive(False)
        
        w=get('entry')
        text=w.get_text()

        if len(text)==0:
            self._get_out()
            return

        if api == GETGLUE:
            params={
                'app':"Rhythmbox",
                'source':"https://github.com/aliva/rhythmbox-microblogger",
                'objectId':'recording_artists/'+self.artist.lower().replace(' ', '_'),
                'comment':(self.title+" from "+self.album)[:140]
            }
            consumer = oauth.Consumer(decode(api['key']), decode(api['secret']))
            token = oauth.Token(decode(conf['token_key']), decode(conf['token_secret']))
            client = oauth.Client(consumer, token, proxy_info=proxy_info)
            link=api['post']+urllib.urlencode(params)

            try:
                resp, content = client.request(link, 'GET')
            except Exception as err:
                w=get('alias')
                w.set_text('Err: %s' % err)
                self._get_out()
                return
       
            if resp['status'] != '200':
                w=get('alias')
                doc=libxml2.parseDoc(content)
                code=doc.xpathEval("/adaptiveblue/error/code")[0].content
                if code=='302':
                    err='Artist not found'
                else:
                    err=code
                w.set_text(' ERR: %s' % err)
                self._get_out()
                return
        else:
            params = {
                'oauth_consumer_key' : decode(api['key']),
                'oauth_signature_method' : 'HMAC-SHA1',
                'oauth_timestamp' : str(int(time.time())),
                'oauth_nonce' : str(random.getrandbits(64)),
                'oauth_version' : '1.0',
                'oauth_token' : decode(conf['token_key']),
            }

            params['status'] = urllib.quote(text, '')
            params['oauth_signature'] = hmac.new(
                '%s&%s' % (decode(api['secret']), decode(conf['token_secret'])),
                '&'.join([
                    'POST',
                    urllib.quote(api['post'], ''),
                    urllib.quote('&'.join(['%s=%s' % (x, params[x])
                                           for x in sorted(params)]), '')
                    ]),
                hashlib.sha1).digest().encode('base64').strip()
            del params['status']

            # post with oauth token
            req = urllib2.Request(api['post'], data = urllib.urlencode(params))
            req.add_data(urllib.urlencode({'status' : text}))
            req.add_header('Authorization', 'OAuth %s' % ', '.join(
                ['%s="%s"' % (x, urllib.quote(params[x], '')) for x in params]))

            try:
                res = urllib2.urlopen(req)
            except Exception as err:
                w=get('alias')
                w.set_text('Err: %s' % err)
                self._get_out()
                return
       
        notif=pynotify.Notification('Message sent to %s' % conf['alias'],
                                    'rbmb',
                                    self.mb.find_file('icon/%s.png' % conf['type']))
        notif.show()
        
        self._get_out()
        w=get('general')
        w.hide_all()

    def _get_out(self):
        self.get('general').set_sensitive(True)
        self.mb.sending=False
        

def decode(string):
    return base64.b64decode(string)
    
def encode(string):
    return base64.b64encode(string)


rbmbSettings.py
#! /usr/bin/python2
# -*- coding: utf8 -*-
#
# Rhythmbox-Microblogger - <http://github.com/aliva/Rhythmbox-Microblogger>
# Copyright (C) 2010 Ali Vakilzade <ali.vakilzade in Gmail>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

__auther__ ='Ali Vakilzade'
__name__   ='rhythmbox-microblogger'
__version__='0.5.2'

import base64
import gconf
import httplib2

KEYS={
    'version'     :'/apps/rhythmbox/plugins/%s/version'      % __name__,
    'template'    :'/apps/rhythmbox/plugins/%s/template'     % __name__,
    'progress'    :'/apps/rhythmbox/plugins/%s/progress'     % __name__,
    'a_list'      :'/apps/rhythmbox/plugins/%s/accounts'     % __name__,
    'a'           :'/apps/rhythmbox/plugins/%s/account/'     % __name__,
    'proxy'       :'/apps/rhythmbox/plugins/%s/proxy'        % __name__,
    'proxy_server':'/apps/rhythmbox/plugins/%s/proxy_server' % __name__,
    'proxy_port'  :'/apps/rhythmbox/plugins/%s/proxy_port'   % __name__,
}

DEFAULT={
    'template'    :'[Rhythmbox] {title} by #{artist} from #{album}',
    'a_list'      :[],
    'progress'    :True,
    'proxy'       :'none',
    'proxy_server':'127.0.0.1',
    'proxy_port'  :8080,
}

class Settings:
    def __init__(self):
        #self._remove_conf(None)
        self._conf=self._read_conf()
        if self._conf is None:
            self._conf=self._create_conf()
        
    def __del__(self):
        del self._conf
        
    def _read_conf(self):
        conf={}

        client=gconf.client_get_default()
        if client.get_string(KEYS['version'])==None:
            return None
        
        ver=client.get_string(KEYS['version'])

        ver=ver.split('.')
        ver[0], ver[1], ver[2]=int(ver[0]), int(ver[1]), int(ver[2])
        if ver[1]<5:
            self._remove_conf(None)
            return None

        if ver[0]==0:
            if ver[1]==5:
                if ver[2]==0:
                    client.set_string(KEYS['template'], DEFAULT['template'])
                if ver[2]<=1:
                    client.set_bool(KEYS['progress'], DEFAULT['progress'])
                    client.set_string(KEYS['proxy'], DEFAULT['proxy'])
                    client.set_string(KEYS['proxy_server'], DEFAULT['proxy_server'])
                    client.set_int(KEYS['proxy_port'], DEFAULT['proxy_port'])

        
        client.set_string(KEYS['version'], __version__)
        
        conf['template']    =client.get_string(KEYS['template'])
        conf['proxy']       =client.get_string(KEYS['proxy'])
        conf['proxy_server']=client.get_string(KEYS['proxy_server'])
        conf['progress']    =client.get_bool  (KEYS['progress'])
        conf['proxy_port']  =client.get_int(KEYS['proxy_port'])
        conf['a_list']      =client.get_list  (KEYS['a_list'], gconf.VALUE_STRING)
        
        conf['a']={}
        
        for alias in conf['a_list']:
            conf['a'][alias]={}
            
            ad=KEYS['a']+alias+'/'
        
            conf['a'][alias]['type']=client.get_string(ad + 'type')
            conf['a'][alias]['alias']=client.get_string(ad + 'alias')
            conf['a'][alias]['token_key']=client.get_string(ad + 'token_key')
            conf['a'][alias]['token_secret']=client.get_string(ad + 'token_secret')
            conf['a'][alias]['url']=client.get_string(ad + 'url')
            conf['a'][alias]['maxlen']=client.get_int(ad + 'maxlen')
        
        return conf
    
    def update_conf(self, text, val):
        client=gconf.client_get_default()

        if text=='progress':
            client.set_bool(KEYS['progress'], val)
        elif text in ('proxy', 'proxy_server'):
            client.set_string(KEYS[text], val)
        elif text=='proxy_port':
            client.set_int(KEYS[text], val)
        elif text=='template':
            if len(val)==0:
                val=DEFAULT['template']
            client.set_string(KEYS['template'], val)
        self._conf[text]=val

    def _create_conf(self):
        client=gconf.client_get_default()
        
        client.set_string(KEYS['version'], __version__)
        client.set_string(KEYS['template'], DEFAULT['template'])
        client.set_bool  (KEYS['progress'], DEFAULT['progress'])
        client.set_list  (KEYS['a_list'], gconf.VALUE_STRING, DEFAULT['a_list'])

        return DEFAULT
    
    def remove_account(self, alias):
        self._conf['a_list'].remove(alias)
        del self._conf['a'][alias]

        self._remove_conf(alias)

        client=gconf.client_get_default()
        client.set_list(KEYS['a_list'], gconf.VALUE_STRING, self._conf['a_list'])
            
    def _remove_conf(self, key):
        client=gconf.client_get_default()
        if key==None:
            add='/apps/rhythmbox/plugins/%s' % __name__
        else:
            add=KEYS['a'] + str(key)
        client.recursive_unset(add, gconf.UNSET_INCLUDING_SCHEMA_NAMES)
        client.remove_dir(add)

        if key==None:
            for i in range(100):
                client.remove_dir(add)
                
    def add_account(self,
                    type,
                    alias,
                    token,
                    token_secret,
                    url,
                    maxlen):
        
        self._conf['a_list'].append(alias)
        
        ad=KEYS['a']+alias+'/'
        
        client=gconf.client_get_default()
        client.set_string(ad + 'type'        , type)
        client.set_string(ad + 'alias'       , alias)
        client.set_string(ad + 'token_key'   , token)
        client.set_string(ad + 'token_secret', token_secret)
        client.set_string(ad + 'url'         , url)
        client.set_int   (ad + 'maxlen'      , maxlen)
        client.set_list  (KEYS['a_list'], gconf.VALUE_STRING, self._conf['a_list'])
        
        self.__init__()

    def get_conf(self, key, alias='object'):
        # a means account
        if key=='a':
            return self._conf['a'][alias]
        elif key=='proxy' and alias=='object':
            return self._get_proxy_configs()
        return self._conf[key]

    def _get_proxy_configs(self):
        try:
            import socks
            proxy=self._conf['proxy']
            if proxy=='none':
                return None
            if proxy=='socks5':
                socks=socks.PROXY_TYPE_SOCKS5
            elif proxy=='socks4':
                socks=socks.PROXY_TYPE_SOCKS4
            else:
                socks=socks.PROXY_TYPE_HTTP

            return httplib2.ProxyInfo(proxy_type=socks,
                                      proxy_host=self._conf['proxy_server'],
                                      proxy_port=self._conf['proxy_port'])
            return None
        except ImportError:
            return None


rbmbAddDialog.py
#! /usr/bin/python2
# -*- coding: utf8 -*-
#
# Rhythmbox-Microblogger - <http://github.com/aliva/Rhythmbox-Microblogger>
# Copyright (C) 2010 Ali Vakilzade <ali.vakilzade in Gmail>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import gtk
import threading
from rbmbRequest import AddAccountRequest
    
class AddDialog(AddAccountRequest):
    def __init__(self, dialog, mb):
        AddAccountRequest.__init__(self)
        
        self.pages=None
        self.mb=mb
        
        assistant=gtk.Assistant()
        
        assistant.set_transient_for(dialog)
        assistant.set_position(gtk.WIN_POS_CENTER_ON_PARENT)
        assistant.set_destroy_with_parent(True)
        assistant.set_modal(True)
        
        assistant=self._attach_pages(assistant)
        
        assistant.set_forward_page_func(self._forward_page_func)
        
        assistant.connect('close' , self._signal, 'close')
        assistant.connect('cancel', self._signal, 'cancel')
        assistant.connect('apply' , self._signal, 'apply')
        
        self.assistant=assistant
        
        assistant.show_all()    
    
    def __del__(self):
        del self.pages
        del self.mb
        del self.assistant
    
    def _forward_page_func(self, page):
        # TODO
        # Update this function
        if self.pages==None:
            return -1

        page_id=self._get_page_id(page)
        type=self._get_account_type()
        self.request_set_type(type)
        find=self._find_page_index
        
        if   page_id=='select':
            return find('authorize')
        elif page_id=='authorize':
            if type=='getglue':
                return find('exchange')
            return find('pin')
        elif page_id=='pin':
            return find('exchange')
        elif page_id=='exchange':
            return find('alias')
        elif page_id=='alias':
            return find('finish')
        return -1
      
    def _signal(self, assistant, id):
        if id=='cancel':
            assistant.destroy()
        elif id=='close':
            self.save_account(self.mb)
            self.mb.config_dialog.update_list_store()
            assistant.destroy()
            
        self.pages=None
        
    def _set_hint(self, hint):
        index=self.assistant.get_current_page()
        
        for page in self.pages:
            if page[1]==index:
                w=page[6].get_object('hint')
                
                try:
                    w.set_text(hint)
                except Exception as err:
                    print 'Could not change hint :\n', err
            
    def _attach_pages(self, assistant):
        pages=[
            # 0 - widget
            # 1 - index
            # 2 - title
            # 3 - type
            # 4 - complete
            # 5 - id
            # 6 - ui
            [None, -1 , 'Select a service provider',
             gtk.ASSISTANT_PAGE_CONTENT, True, 'select', None],
            [None, -1 , 'Authorize account',
             gtk.ASSISTANT_PAGE_CONTENT, False, 'authorize', None],
            [None, -1 , 'Pin Code',
             gtk.ASSISTANT_PAGE_CONTENT, False, 'pin', None],
            [None, -1 , 'Exchange Request',
             gtk.ASSISTANT_PAGE_CONTENT, False, 'exchange', None],
            [None, -1 , 'Basic auth',
             gtk.ASSISTANT_PAGE_CONTENT, False, 'basic', None],
            [None, -1 , 'Alias',
             gtk.ASSISTANT_PAGE_CONTENT, False, 'alias', None],
            [None, -1 , 'Finished',
             gtk.ASSISTANT_PAGE_CONFIRM, True, 'finish', None],
        ]
        
        for page in pages:
            page[6]=gtk.Builder()
            page[6].add_from_file(self.mb.find_file('ui/rbmb-add-'+page[5]+'.ui'))
            page[6].connect_signals(self)
        
            page[0]=page[6].get_object('general')
            
            page[1]=assistant.append_page(page[0])
            
            assistant.set_page_title   (page[0], page[2])
            assistant.set_page_type    (page[0], page[3])
            assistant.set_page_complete(page[0], page[4])
            
            if page[5]=='select':
                self._set_service_logo(page[6])
            
        
        self.pages=pages
        return assistant
    
    def _set_service_logo(self, ui):
        for account in self.mb.ACCOUNT_TYPE:
            w=ui.get_object(account+'-image')
            w.set_from_stock('rbmb-'+account, gtk.ICON_SIZE_MENU)
            
    def _get_account_type(self):
        ui=self.pages[0][6]
        
        for account in self.mb.ACCOUNT_TYPE:
            w=ui.get_object(account)
            if w.get_active():
                return account
            
    def _find_page_index(self, id):
        for page in self.pages:
            if page[5]==id:
                return page[1]
        return -1

    def _get_page_id(self, index):
        for page in self.pages:
            if page[1]==index:
                return page[5]
        return None
    
    def _authorize_clicked(self, button):
        page=self.pages[1]
                    
        self._set_hint('')
        
        threading.Thread(target=self.authorize,
                args=(self._set_hint, button, self.assistant, page, self.mb.get_conf('proxy'))
                ).start()
        
    def _exchange_clicked(self, button):
        page=self.pages[3]        
        
        self._set_hint('')
        
        threading.Thread(target=self.exchange,
                args=(self._set_hint, button, self.assistant, page, self.mb.get_conf('proxy'))
                ).start()
        
    def _pin_entry_changed(self, entry):
        len=entry.get_text_length()
        self.pin=entry.get_text()
        
        page=self.pages[2]
        page[4]=bool(len)
        self.assistant.set_page_complete(page[0], page[4])
        
    def _alias_entry_changed(self, entry):
        len=entry.get_text_length()
        self.alias=entry.get_text().strip()
        
        self._set_hint('')
        
        page=self.pages[5]
        
        if self.alias in self.mb.get_conf('a_list'):
            page[4]=False
            self.assistant.set_page_complete(page[0], page[4])
            self._set_hint('Use another alias\n%s is not unique' % self.alias)
            return
        
        page[4]=bool(len)
        self.assistant.set_page_complete(page[0], page[4])


rhythmbox-microblogger.py
#! /usr/bin/python2
# -*- coding: utf8 -*-
#
# Rhythmbox-Microblogger - <http://github.com/aliva/Rhythmbox-Microblogger>
# Copyright (C) 2010 Ali Vakilzade <ali.vakilzade in Gmail>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


__auther__ ='Ali Vakilzade'
__name__   ='rhythmbox-microblogger'
from rbmbSettings import __version__

import gtk
import pynotify
import rb
import rhythmdb
import threading
import urllib
from rbmbConfigDialog import ConfigDialog
from rbmbSettings import Settings
from rbmbSettings import DEFAULT
from rbmbRequest import Post

UI_TOOLBAR='''
<ui>
    <toolbar name='ToolBar'>
        <placeholder name='PluginPlaceholder'>
            <toolitem name='rbmb-toolitem-%s' action='rbmb-action-%s'/>
        </placeholder>
    </toolbar>
</ui>
'''

class microblogger(rb.Plugin):
    ACCOUNT_TYPE=('twitter', 'identica', 'getglue')
    def __init__(self):
        rb.Plugin.__init__(self)

    def activate(self, shell):
        gtk.gdk.threads_init()
        pynotify.init('Rhythmbox')
        
        self.shell=shell
        self.pl=shell.get_property('shell-player')
        self.db=shell.get_property('db')

        
        self.config_dialog=ConfigDialog(self)
        self.settings=Settings()
        self.post=Post(self)
        
        self.get_conf=self.settings.get_conf

        self.uim = shell.get_ui_manager()
                        
        self._register_icons()
        
        self.add_ui()
        self._attach_box()
        
        self.sending=False
        
        threading.Thread(target=self._check_4_update).start()

    def deactivate(self, shell):
        self.remove_ui()
        
        pynotify.uninit()
        
        del self.boxui
        del self.config_dialog
        del self.settings
    
    def create_configure_dialog(self, dialog=None):
        if dialog==None:
            dialog=self.config_dialog.create_main_window()
        return dialog

    def _register_icons(self):
        IconFactory=gtk.IconFactory()
        IconFactory.add_default()

        for account in self.ACCOUNT_TYPE:
            gtk.stock_add([('rbmb-%s' % account, account, 0, 0, '')])
            IconSource=gtk.IconSource()
            IconSet=gtk.IconSet()
            IconSource.set_filename(self.find_file('icon/%s.png' % account))
            IconSet.add_source(IconSource)
            IconFactory.add('rbmb-%s' % account, IconSet)
            
    def add_ui(self):
        self.ui_id=[]
        self.action_groups=[]

        for alias in self.get_conf('a_list'):
            action=gtk.Action('rbmb-action-%s' % alias,
                              alias,
                              'Microblogger plugin',
                              'rbmb-%s' % self.get_conf('a', alias)['type'])
            activate_id = action.connect('activate', self._send_clicked, alias)
            action_group = gtk.ActionGroup('rbmb-actiongroup-%s'% alias)
            action_group.add_action(action)
            self.uim.insert_action_group(action_group, 0)

            self.ui_id.append(self.uim.add_ui_from_string(UI_TOOLBAR % (alias, alias)))
            self.action_groups.append(action_group)

        self.uim.ensure_update()
        
    def remove_ui(self):
        for key in self.ui_id:
            self.uim.remove_ui(key)

        for key in self.action_groups:
            self.uim.remove_action_group(key)

        del self.ui_id
        del self.action_groups

        self.uim.ensure_update()
    
    def _attach_box(self):
        self.boxui=gtk.Builder()
        self.boxui.add_from_file(self.find_file('ui/rbmb-box.ui'))
        self.boxui.connect_signals(self)
        
        box=self.boxui.get_object('general')

        self.shell.add_widget (box, rb.SHELL_UI_LOCATION_MAIN_TOP)
        
        w=self.boxui.get_object('cancel-image')
        w.set_from_stock(gtk.STOCK_CANCEL, gtk.ICON_SIZE_BUTTON)
		
        box.hide_all()
        
    def _send_clicked(self, button, alias):
        if self.sending:
            return
           
        box=self.boxui.get_object('general')
        box.set_sensitive(True)
        box.hide_all()
        
        if not self.pl.get_playing():
            return
     
        self.alias=alias
        
        get=self.boxui.get_object
        
        conf=self.get_conf('a', alias)
        
        self.pl_entry=self.pl.get_playing_entry()

        w=get('type')
        w.set_from_stock('rbmb-'+ conf['type'], gtk.ICON_SIZE_BUTTON)
        
        w=get('alias')
        w.set_text(alias)
        
        w=get('len')
        w.set_text('')
        
        w=get('entry')
        w.set_progress_fraction(0)
        w.set_sensitive(True)
        w.set_text(self.get_conf('template'))
        w.grab_focus()
        
        w=get('send')
        w.set_sensitive(True)

        box.show_all()

        if conf['type']=='getglue':
            w=get('entry')
            w.set_sensitive(False)
            w.set_text('  ')
        
    def _send_thread(self, button):
        self.sending=True
        
        threading.Thread(target=self.post.post,
                         args=(self.boxui,
                               self.artist, self.title, self.album,
                               self.alias, self.get_conf('proxy'))
                         ).start()
        
    def _cancel_clicked(self, button):
        self.sending=False
        
        box=self.boxui.get_object('general')
        box.hide_all()
        
    def _entry_changed(self, entry):
        maxlen=140
        
        len=entry.get_text_length()
        
        label=self.boxui.get_object('len')
        label.set_text('%d' % (maxlen-len))
        
        if self.get_conf('progress'):
            entry.set_progress_fraction(float(len)/maxlen)
        
        send=self.boxui.get_object('send')
        send.set_sensitive(0<len<=maxlen)
        
        text=entry.get_text()
        pl_entry=self.pl_entry
        if pl_entry==None:
            return

        get=self.db.entry_get
        items={
            '{title}' :get(pl_entry, rhythmdb.PROP_TITLE),
            '{genre}' :get(pl_entry, rhythmdb.PROP_GENRE),
            '{artist}':get(pl_entry, rhythmdb.PROP_ARTIST),
            '{album}' :get(pl_entry, rhythmdb.PROP_ALBUM),
            '{rate}'  :get(pl_entry, rhythmdb.PROP_RATING),
            '{year}'  :get(pl_entry, rhythmdb.PROP_YEAR),
            '{pcount}':get(pl_entry, rhythmdb.PROP_PLAY_COUNT),
        }
        self.artist = get(pl_entry, rhythmdb.PROP_ARTIST)
        self.title = get(pl_entry, rhythmdb.PROP_TITLE)
        self.album = get(pl_entry, rhythmdb.PROP_ALBUM)
        del get
        
        for item in items:
            # if has tag
            tmp=str(items[item])
            tmp=self._remove_forbidden_chars(tmp)
            text=text.replace('#'+item, '#'+tmp)

            # if hasn't tag
            text=text.replace(item, str(items[item]))

        entry.set_text(text)

    def _entry_key_press_event(self, entry, event):
        if gtk.gdk.keyval_name(event.keyval)=='Escape':
            self._cancel_clicked(None)

    def _remove_forbidden_chars(self, text):
        chars='!@#$%^&*()-+=\\|/[]{};:\'"<>,? '

        for c in chars:
            text=text.replace(c, '_')

        for c in ('____', '___', '__'):
            text=text.replace(c, '_')

        return text.strip(chars+'_')
    
    def _entry_icon_press(self, entry, pos, event): 
        entry.set_text('')
        
    def _check_4_update(self):        
        print 'Checking for updates...'
        
        try:
            link=urllib.urlopen('https://github.com/downloads/aliva/rhythmbox-microblogger/VERSION')
        except Exception as err:
            print 'Coudn\'t get update info from rbmb'
            return      
        if link.code!=200:
            print 'Coudn\'t get update info from rbmb'
            return
        
        f=link.read()
        link.close()
        
        f=f.split('\n')[0]
        f=f.split('=')
        
        if f[0]!='version':
            print 'Errors in update info file'
            return
        
        f=f[1]
        f=f.split('.')
        f=[int(x) for x in f]
        

        old=False
       
        ver=__version__.split('.')
        ver=[int(x) for x in ver]
        
        if   f[0]>ver[0]:
            old=True
        elif f[0]==ver[0] and f[1]>ver[1]:
            old=True
        elif f[0]==ver[0] and f[1]==ver[1] and f[2]>ver[2]:
            old=True

        if not old:
            return

        notif=pynotify.Notification('Updates available for MicroBlogger Plugin',
                                    'version %d.%d.%d is ready!' % (f[0], f[1], f[2]))
        notif.show()
